### PR TITLE
PHP7: fix ubuntu compile error

### LINKED ISSUE #7642: php grpc ext fails to compile (php 7.0.8, ubuntu 16.04)
>  `$
> 
>  make       
> /bin/bash /var/www/grpc/src/php/ext/grpc/libtool --mode=compile cc -Wall -Werror -std=c11 -I. -I/var/www/grpc/src/php/ext/grpc -DPHP_ATOM_INC -I/var/www/grpc/src/php/ext/grpc/include -I/var/www/grpc/src/php/ext/grpc/main -I/var/www/grpc/src/php/ext/grpc -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib -I/usr/local/include  -DHAVE_CONFIG_H  -g -O2   -c /var/www/grpc/src/php/ext/grpc/call.c -o call.lo 
> libtool: compile:  cc -Wall -Werror -std=c11 -I. -I/var/www/grpc/src/php/ext/grpc -DPHP_ATOM_INC -I/var/www/grpc/src/php/ext/grpc/include -I/var/www/grpc/src/php/ext/grpc/main -I/var/www/grpc/src/php/ext/grpc -I/usr/include/php/20151012 -I/usr/include/php/20151012/main -I/usr/include/php/20151012/TSRM -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/ext -I/usr/include/php/20151012/ext/date/lib -I/usr/local/include -DHAVE_CONFIG_H -g -O2 -c /var/www/grpc/src/php/ext/grpc/call.c  -fPIC -DPIC -o .libs/call.o
> /var/www/grpc/src/php/ext/grpc/call.c: In function ‘grpc_parse_metadata_array’:
> cc1: error: function may return address of local variable [-Werror=return-local-addr]
> In file included from /var/www/grpc/src/php/ext/grpc/php_grpc.h:60:0,
>                  from /var/www/grpc/src/php/ext/grpc/call.h:44,
>                  from /var/www/grpc/src/php/ext/grpc/call.c:34:
> /var/www/grpc/src/php/ext/grpc/php7_wrapper.h:146:8: note: declared here
>    zval _stack_zval_##pzv; \
>         ^
> /var/www/grpc/src/php/ext/grpc/call.c:89:3: note: in expansion of macro ‘PHP_GRPC_MAKE_STD_ZVAL’
>    PHP_GRPC_MAKE_STD_ZVAL(array);
>    ^
> cc1: error: function may return address of local variable [-Werror=return-local-addr]
> /var/www/grpc/src/php/ext/grpc/php7_wrapper.h:146:8: note: declared here
>    zval _stack_zval_##pzv; \
>         ^
> /var/www/grpc/src/php/ext/grpc/call.c:89:3: note: in expansion of macro ‘PHP_GRPC_MAKE_STD_ZVAL’
>    PHP_GRPC_MAKE_STD_ZVAL(array);
>    ^
> /var/www/grpc/src/php/ext/grpc/call.c: In function ‘grpc_php_wrap_call’:
> /var/www/grpc/src/php/ext/grpc/call.c:190:10: error: function returns address of local variable [-Werror=return-local-addr]
>    return call_object;
>           ^
> In file included from /var/www/grpc/src/php/ext/grpc/php_grpc.h:60:0,
>                  from /var/www/grpc/src/php/ext/grpc/call.h:44,
>                  from /var/www/grpc/src/php/ext/grpc/call.c:34:
> /var/www/grpc/src/php/ext/grpc/php7_wrapper.h:146:8: note: declared here
>    zval _stack_zval_##pzv; \
>         ^
> /var/www/grpc/src/php/ext/grpc/call.c:185:3: note: in expansion of macro ‘PHP_GRPC_MAKE_STD_ZVAL’
>    PHP_GRPC_MAKE_STD_ZVAL(call_object);
>    ^
> cc1: all warnings being treated as errors
> Makefile:196: recipe for target 'call.lo' failed
> make: *** [call.lo] Error 1
> 
> `

I tried `v1.0.x`, same result

### PR DESCRIPTION
**Pull request by stanley-cheung**
Fixes #7642

### CODE REVIEW

**Comment by murgatroid99** [comment]
That macro is used inside of functions. I don't think making the variable static is a good idea.

**Comment by stanley-cheung** [comment]
Yea I agree this isn't ideal. But as it stands now, the extension isn't installable on Ubuntu. A rewrite will be much more involved. Can we unblock this for now, and rewrite it in maybe 1.0.1?

**Comment by pgrosu** [comment]
@murgatroid99 For now something is better than nothing, and let's just get as rapidly as possible to a 1.0 production release, which I'm getting this strong feeling is the highest priority for this project.  Since gRPC is a quick Stubby/ARCWire/Sake integration-rewrite, let's wait until the [developer docs](https://github.com/grpc/grpc/issues/7636) come out, and then we can take our time by carefully going as a community through the reasoning behind the design, analyzing all the data flow possibilities, and propose places where optimizations in implementation might become beneficial as part of the 1.1 release.

Hope you don't mind,
~p

**Comment by murgatroid99** [comment]
LGTM

**Comment by nicolasnoble** [comment]
@pgrosu the original commit wasn't going to work _at all_. Polishing and optimizing things after the release is one thing, but committing broken code just for the sake of having a test pass somewhere isn't something we should even question.

The new fix seems indeed better.

**Comment by stanley-cheung** [comment]
Yea sorry about that. I completely misunderstood the problem previously

**Comment by pgrosu** [comment]
@nicolasnoble I agree that, that is a given :)  I assumed the code somewhere is already performing the memory allocation automatically for a `_zval_struct`, as he was just passing the address and the PHP tests were passing in Jenkins.  

It might be good to have some helper functions, so that things like this don't surprise us in the future.  I proposed to in https://github.com/grpc/grpc-java/pull/2137#discussion_r73775688 to create an `Optional` type so that things are easier to manage, and maybe the same can be performed here where `PHP_GRPC_MAKE_STD_ZVAL` is defined to create and allocate at the same time.  Basically the less a developer has at to worry about, the more they can focus on the design - which is already quite complex, when one has to walk through all the intricacies of the code-base.

File: src/php/ext/grpc/php7_wrapper.h [hunk 1]
```diff
@@ -143,8 +143,7 @@ static inline int php_grpc_zend_hash_find(HashTable *ht, char *key, int len,
 
 #define PHP_GRPC_RETURN_STRING(val, dup) RETURN_STRING(val)
 #define PHP_GRPC_MAKE_STD_ZVAL(pzv) \
-  zval _stack_zval_##pzv; \
-  pzv = &(_stack_zval_##pzv)
+  pzv = (zval *)emalloc(sizeof(zval));
```
Comment (thinkerou) [inline src/php/ext/grpc/php7_wrapper.h]: The end of function should have `efree`?

Comment (stanley-cheung) [inline src/php/ext/grpc/php7_wrapper.h]: I believe that will be taken care of in `RETURN_DESTROY_ZVAL`. We can certainly do an audit to make sure there's no other leak.

**Pull request merged**